### ISSUE TITLE
The result seems okay, but there are a few confusing things

### ISSUE THREAD

**Issue by hoangcuong2011** [idiap/importance-sampling#17]
Hi,

I happened to cross by the Importance Sampling topic and I crossed by your project. Thanks for providing us a great project so we can give a try on this idea.

I modified a very standard code for mnist classification from this link (Note I also drop the Dropout layer for a reason I will explain later): https://github.com/keras-team/keras/blob/master/examples/mnist_cnn.py

I noticed the result seems okay, at least using Importance Sampling seems gives good result faster. But it is not so clear whether it improves the result, though!

Baseline (Note I drop the Dropout layer in the baseline as well):
Epoch 1 - val_accuracy: 0.9815
Epoch 2 - val_accuracy: 0.9839
Epoch 3 - val_accuracy: 0.9897
Epoch 4 - val_accuracy: 0.9916
Epoch 5 - val_accuracy: 0.9897

Using Importance Sampling
Epoch 1 - val_accuracy: 0.9828
Epoch 2 - val_accuracy: 0.9893
Epoch 3 - val_accuracy: 0.9874
Epoch 4 - val_accuracy: 0.9902
Epoch 5 - val_accuracy: 0.9897

I wanted to ask three questions:
1. Did you observe significant improvements when using Importance Sampling than NOT using Importance Sampling? Or the main goal of using Importance Sampling is rather getting a good result faster (at least faster in terms of training with a smaller number of epochs)
2. Training with Importance Sampling seems pretty slow (running time for each epoch), actually. Is there any chance to speed this up?
3. Finally, I noticed we should take care of using Dropout and BatchNormalization in Importance Sampling, as: "Those layers may affect the importance calculations and you are advised to exchange them for LayerNormalization or BatchNormalization in test mode and L2 regularization." Why is it the case?

Many thanks in advance.

Best,

---
just in case: My code for training with Importance Sampling
```

import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras import backend as K
from importance_sampling.training import ImportanceTraining
from keras.layers import Dense, Activation

batch_size = 128
num_classes = 10
epochs = 12

# input image dimensions
img_rows, img_cols = 28, 28

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dense(num_classes))
model.add(Activation("softmax"))


model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

ImportanceTraining(model).fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test)
)
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

```

**Comment by angeloskath** [comment]
Hi,

Thanks for using the project I 'll try to answer the questions in turn:

1. The goal of the project is getting results faster in terms of wall clock time. You either need to train less epochs or train with the ConstantTimeImportanceTraining wrapper to achieve that. Since we create an unbiased estimator it is not expected to get better results, however if the classes are unbalanced or there is some other reason for high variance that doesn't allow the training to converge one could see better results (increasing the batch size would have the same effect). Using your code, indeed the test error is not better however, the training loss at **epoch 7** is like the one achieved at **epoch 12** with SGD which means that it might take twice as much to train with IS but you need half the training.
2. Yes you could use the [ConstantTimeImportanceTraining](https://github.com/idiap/importance-sampling/blob/991230a86d6d3d3a1d8b1a84716ae0db92248e26/importance_sampling/training.py#L403) which for large models should have negligible impact on training time per epoch. I should add this to the documentation (PRs are welcome).
3. This can be fixed but not automatically by the library. Let me explain the way IS works (assuming we are using IS because the variance can be reduced see the [paper](https://arxiv.org/pdf/1803.00942.pdf) for details):
    &nbsp; a. We perform a forward pass for N samples
    &nbsp; b. We compute the importance distribution based on the forward pass
    &nbsp; c. We sample n samples from the N using the importance distribution
    &nbsp; d. We train on the n using weights to make the estimator unbiased
When using stochastic layers, such as Dropout or BatchNorm, the importance distribution changes based on the specific instantiation of the random variables in those stochastic layers (dropout mask and mean, variance). One could fix those layers from step a to d and then they could be used without any problem. Simpler solution is to change those layers with deterministic ones like GroupNorm or fix the statistics after some epochs.

Perhaps, I should provide Dropout and BatchNorm layers that automatically work with IS...

For more examples of IS you can see the [examples](http://idiap.ch/~katharas/importance-sampling/examples/) in the documentation.

**Comment by hoangcuong2011** [comment]
Thank you so much for your detailed answer @angeloskath !
Have a good day!
